### PR TITLE
Initial version of MapReduce guide

### PR DESCRIPTION
**Pull request by sreevatsanraman**
This PR contains
- Code for MapReduce guide
- Unit test 
- Packaging into ZIP 
- README updates

### CODE REVIEW

File: src/main/assembly/assembly.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 http://maven.apache.org/xsd/assembly-1.1.2.xsd">
+  <id>${project.version}</id>
+  <formats>
+    <format>zip</format>
+  </formats>
+  <includeBaseDirectory>false</includeBaseDirectory>
+  <fileSets>
+    <fileSet>
+      <directory>${sdk.dir}</directory>
```
Comment (abaranau) [inline src/main/assembly/assembly.xml]: sdk.dir?

File: src/main/assembly/assembly.xml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<!--
```
Comment (abaranau) [inline src/main/assembly/assembly.xml]: should we remove for now, since we are not pacakging?

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+  License for the specific language governing permissions and limitations under
+  the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>co.cdap.guides</groupId>
```
Comment (abaranau) [inline pom.xml:21]: co.cask.cdap

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+      <version>${cdap.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>${slf4j.version}</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
```
Comment (abaranau) [inline pom.xml:62]: move test dependencies to the end

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+          <groupId>org.slf4j</groupId>
+          <artifactId>slf4j-log4j12</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-mapreduce-client-core</artifactId>
+      <version>${hadoop.version}</version>
+      <exclusions>
```
Comment (abaranau) [inline pom.xml:99]: should we just say it is provided and not worry about exclusions?

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+          <groupId>org.slf4j</groupId>
+          <artifactId>slf4j-log4j12</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop.version}</version>
+      <exclusions>
```
Comment (abaranau) [inline pom.xml]: same

Comment (abaranau) [inline pom.xml]: oh, you need it for tests??

File: pom.xml [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        </exclusion>
+        <exclusion>
+          <artifactId>jsp-api</artifactId>
+          <groupId>javax.servlet.jsp</groupId>
+        </exclusion>
+        <exclusion>
+          <artifactId>slf4j-api</artifactId>
+          <groupId>org.slf4j</groupId>
+        </exclusion>
+      </exclusions>
```
Comment (abaranau) [inline pom.xml]: this exclusion list looks ugly

Comment (abaranau) [inline pom.xml]: see https://github.com/caskdata/cdap-apps/blob/develop/Wise/pom.xml

File: src/main/java/co/cdap/guides/ClientCount.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cdap.guides;
+
+import com.google.common.base.Objects;
+
+/**
+ * Pojo to store clientIP and count.
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/ClientCount.java]: Holds clientIP, count pairs

File: src/main/java/co/cdap/guides/ClientCount.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  public String getClientIP() {
+    return clientIP;
+  }
+
+  public int getCount() {
+    return count;
+  }
+
+  @Override
+  public String toString() {
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/ClientCount.java]: you need those?

File: src/main/java/co/cdap/guides/ClientCount.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    return Objects.toStringHelper(this)
+                    .add("ip", clientIP)
+                    .add("count", count)
+                    .toString();
+  }
+
+  @Override
+  public int compareTo(ClientCount clientCount) {
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/ClientCount.java:40]: and this?

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public class LogAnalyticsApp extends AbstractApplication {
+
+  public static final String DATASET_NAME = "resultStore";
+  public static final byte [] DATASET_RESULTS_KEY = {'r'};
+
+  @Override
+  public void configure() {
+    setName("LogAnalyticsApp");
+    setDescription("An application that computes top 10 clientIPs from Apache access log data");
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java]: I'd remove

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    addStream(new Stream("logEvent"));
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java]: logEvents

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import com.google.inject.util.Types;
+
+
+/**
+ * LogAnalyticsApp that consumes Apache access log event, computes topN clientIps by traffic and a service
+ * to retrieve the results.
+ */
+public class LogAnalyticsApp extends AbstractApplication {
+
+  public static final String DATASET_NAME = "resultStore";
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java]: tell which dataset. RESULTS_DATASET_NAME? or TOPCLIENTS_DATASET_NAME?

File: src/main/java/co/cdap/guides/TopClientsService.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+  @Override
+  protected void configure() {
+    setName("TopClientsService");
+    addHandler(new ResultsHandler());
+  }
+
+  public static class ResultsHandler extends AbstractHttpServiceHandler {
+
+    @UseDataSet(LogAnalyticsApp.DATASET_NAME)
+    private ObjectStore<List<ClientCount>> topN;
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/TopClientsService.java]: topClients?

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+  @Override
+  public void configure() {
+    setName("LogAnalyticsApp");
+    setDescription("An application that computes top 10 clientIPs from Apache access log data");
+    addStream(new Stream("logEvent"));
+    addMapReduce(new TopClientsMapReduce());
+    addService(new TopClientsService());
+    try {
+      DatasetProperties props = ObjectStores.objectStoreProperties(Types.listOf(ClientCount.class),
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java:43]: using objectStore seems confusing... Let's think

File: src/main/java/co/cdap/guides/ClientCount.java (hunk 8 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/ClientCount.java]: Fixed.

File: src/main/java/co/cdap/guides/ClientCount.java (hunk 9 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/ClientCount.java]: No using it. Will remove.

File: src/main/java/co/cdap/guides/ClientCount.java (hunk 10 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/ClientCount.java:40]: I need this, Priority queue to get topN

File: src/main/java/co/cdap/guides/package-info.java [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * Package CDAP MapReduce guide.
+ */
+package co.cdap.guides;
```
Comment (abaranau) [inline src/main/java/co/cdap/guides/package-info.java]: rm file. doesn't add value

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java (hunk 11 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java]: Done

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java (hunk 13 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java]: Used RESULTS_DATASET_NAME

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java (hunk 12 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java]: Fixed.

File: README.rst [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+===============================
+
+`MapReduce <http://research.google.com/archive/mapreduce.html>`_ is the most popular paradigm for processing large
+amounts of data in a reliable and fault-tolerant manner. In this guide you will learn how to batch process data using
+MapReduce in the `Cask Data Application Platform (CDAP). <http://cdap.io>`_
+
+What You Will Build
+-------------------
+
+This guide will take you through building a CDAP application that ingests Apache access log events, computes
```
Comment (abaranau) [inline README.rst]: that uses ingested Apache access log events to compute top 10...

File: src/main/java/co/cdap/guides/LogAnalyticsApp.java (hunk 15 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/LogAnalyticsApp.java:43]: Other options are to use custom dataset or use table and encode the top results. Both of which will make the example more complex. open to ideas.

File: src/main/java/co/cdap/guides/TopClientsService.java (hunk 14 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/TopClientsService.java]: Fixed.

File: README.rst [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+top 10 client IPs that sends requests in the last hour and query the results. You will:
+
+* Build MapReduce program to process events from Apache access log
```
Comment (abaranau) [inline README.rst]: MapReduce program - Mapreduce job. Also link it to the MR section in the guide

File: README.rst [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+* Use `Dataset <http://docs.cask.co/cdap/current/en/dev-guide.html#datasets>`_ to persist the result of a
```
Comment (abaranau) [inline README.rst]: results?

File: src/main/assembly/assembly.xml (hunk 2 above)
Comment (sreevatsanraman) [inline src/main/assembly/assembly.xml]: Done.

File: README.rst [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+------------------
+
+* `JDK 6 or JDK 7 <http://www.oracle.com/technetwork/java/javase/downloads/index.html>`_
+* `Apache Maven 3.0+ <http://maven.apache.org/>`_
+* `CDAP SDK <http://docs.cdap.io/cdap/current/en/getstarted.html#download-and-setup>`_
+
+Let’s Build It!
+---------------
+
+Following sections will guide you through building an application from scratch. You can download its source code and binaries from `here <placeholder..>`_
```
Comment (abaranau) [inline README.rst]: clone from this github repo. You can skip ..  - use text from other guides

File: README.rst [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+
+Let’s Build It!
+---------------
+
+Following sections will guide you through building an application from scratch. You can download its source code and binaries from `here <placeholder..>`_ 
+
+Application Design
+------------------
+
+The application will assume that we are processing Apache access log events from Apache servers. The log events can be ingested
```
Comment (abaranau) [inline README.rst]: In this guide we assume that Appache access logs are aggregated and ingested into a Stream. The log events can be ingested ....

File: README.rst [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+into a Stream continuously in real-time or in batches, which doesn’t affect the ability to consume it by MapReduce.
+
+MapReduce job extract necessary information from raw logs and computes top 10 Client IPs by traffic in a specific time range.
```
Comment (abaranau) [inline README.rst]: extracts

File: README.rst [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+
+
+Implementation
+--------------
+
+The first step is to get our application structure set up.  We will use a standard Maven project structure for all of
+the source code files::
+
+  ./pom.xml
+  ./README.rst
```
Comment (abaranau) [inline README.rst]: no README here needed - and other stuff that is not required - test, assembly, etc.

File: pom.xml (hunk 7 above)
Comment (sreevatsanraman) [inline pom.xml]: Right. This is cleaner! Fixed.

File: README.rst [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+                                                                     DatasetProperties.EMPTY);
+        createDataset(DATASET_NAME, ObjectStore.class, props);
+      } catch (UnsupportedTypeException e) {
+        throw Throwables.propagate(e);
+      }
+      addService(new TopClientsService());
+    }
+  }
+
+The LogAnalyticsapplication defines a new `Stream <http://docs.cdap.io/cdap/current/en/dev-guide.html#streams>`_ to ingest the Apache access log events.
```
Comment (abaranau) [inline README.rst]: Stream that Apache access logs are ingested to.

File: README.rst [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+The log events can be ingested into CDAP using a RESTful API. Once the data is ingested into the stream the events
```
Comment (abaranau) [inline README.rst]: into CDAP Stream

Comment (abaranau) [inline README.rst]: I'd remove "The log events can be ingested into CDAP using a RESTful API." sentence - it is out of the context. You'll show later stuff "in action"

File: pom.xml (hunk 5 above)
Comment (sreevatsanraman) [inline pom.xml:99]: Fixed. Same as Wise.

File: pom.xml (hunk 6 above)
Comment (sreevatsanraman) [inline pom.xml]: Fixed. Same as Wise.

File: pom.xml (hunk 4 above)
Comment (sreevatsanraman) [inline pom.xml:62]: Done.

File: README.rst [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+The results of the MapReduce job is persisted into a Dataset, the application uses createDataset method to define
+the Dataset. Finally, the application adds a service to query the results from the Dataset.
+
+Let's take a closer look at the MapReduce program.
+
+The TopClientsMapReduce job extends an 
+`AbstractMapReduce <http://docs.cdap.io/cdap/2.5.1/en/javadocs/co/cask/cdap/api/mapreduce/AbstractMapReduce.html>`_
+class and overrides the ``configure()`` and ``beforeSubmit()`` methods.
+
+* ``configure()`` method configures a MapReduce job, by returning an instance of
```
Comment (abaranau) [inline README.rst]: job, by setting job name, description and output Dataset.

Comment (abaranau) [inline README.rst]: that's it, otherwise too many words

File: src/main/java/co/cdap/guides/package-info.java (hunk 16 above)
Comment (sreevatsanraman) [inline src/main/java/co/cdap/guides/package-info.java]: Done.

File: README.rst [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+`MapReduceSpecification <http://docs.cdap.io/cdap/2.5.1/en/javadocs/co/cask/cdap/api/mapreduce/MapReduceSpecification.html>`_. The MapReduce
+job name, description and output Dataset are configured in the example.
+
+* ``beforeSubmit()`` method is invoked at runtime, before the MapReduce job is executed. Here, you will have access to the
```
Comment (abaranau) [inline README.rst]: you can access?

File: README.rst [hunk 28]
```diff
@@ -0,0 +1,272 @@
...
+
+      // Read events from last 60 minutes as input to the mapper.
+      final long endTime = context.getLogicalStartTime();
+      final long startTime = endTime - TimeUnit.MINUTES.toMillis(60);
+      StreamBatchReadable.useStreamInput(context, "logEvent", startTime, endTime);
+    }
+  }
+
+
+The next step is to implement the Mapper and Reduce classes. The Mapper and Reducer classes extend from the standard
```
Comment (abaranau) [inline README.rst]: "The Mapper and Reducer classes extend from..." - "In this example Mapper and Reducer are built by implementing Hadoop APIs".

File: README.rst [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+      String streamBody = value.toString();
+      if (streamBody != null  && !streamBody.isEmpty()) {
+        String ip = streamBody.substring(0, streamBody.indexOf(" "));
+        // Map output Key: IP and Value: Count
+        context.write(new Text(ip), OUTPUT_VALUE);
+      }
+    }
+  }
+
+The reducer class gets the clientIP and count from the map jobs and then aggregates the count for each cilentIP and
```
Comment (abaranau) [inline README.rst:175]: or smth like that

File: README.rst [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+    @Override
+    protected void configure() {
+      setName("TopClientsService");
+      addHandler(new ResultsHandler());
+    }
+
+    public static class ResultsHandler extends AbstractHttpServiceHandler {
+
+      @UseDataSet(LogAnalyticsApp.DATASET_NAME)
+      private ObjectStore<List<ClientCount>> topN;
```
Comment (abaranau) [inline README.rst:199]: topClients

File: README.rst (hunk 23 above)
Comment (sreevatsanraman) [inline README.rst]: Removed README and assembly. Will leave the test.

File: README.rst (hunk 30 above)
Comment (abaranau) [inline README.rst:199]: or topIPs

File: README.rst [hunk 31]
```diff
@@ -0,0 +1,272 @@
...
+        }
+      }
+    }
+  }
+
+
+Build and Run
+-------------
+
+The LogAnalyticsApp can be built and packaged using standard Apache maven command:
```
Comment (abaranau) [inline README.rst]: rm "standard"

File: README.rst [hunk 32]
```diff
@@ -0,0 +1,272 @@
...
+  cdap-cli.sh send stream logEvent "255.255.255.185 - - [23/Sep/2014:11:45:38 -0400] \"GET /coopr.html HTTP/1.0\" 200 121 \" \"Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1)\"\n"
+  cdap-cli.sh send stream logEvent "255.255.255.182 - - [23/Sep/2014:11:45:38 -0400] \"GET /tigon.html HTTP/1.0\" 200 111 \" \"Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1)\"\n"
+  cdap-cli.sh send stream logEvent "255.255.255.182 - - [23/Sep/2014:11:45:38 -0400] \"GET /tigon.html HTTP/1.0\" 200 145 \" \"Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1)\"\n"
+
+
+We can now start the MapReduce job to process the events that were ingested
+
+  cdap-cli.sh start mapreduce LogAnalyticsApp.TopClientsMapReduce
+
+The MapReduce job will take a couple of minutes to process the sample data ingested.
```
Comment (abaranau) [inline README.rst]: to process. &lt;nothing after>

File: README.rst [hunk 33]
```diff
@@ -0,0 +1,272 @@
...
+
+We can now start the TopClients service and check the service calls::
```
Comment (abaranau) [inline README.rst]: "and check the service calls " -> "and query for processing results"

File: README.rst [hunk 34]
```diff
@@ -0,0 +1,272 @@
...
+  placeholder
+
+You have now learnt how to write MapReduce job to process events from a stream, write results to a DataSet and query
+the results using services.
+
+Extend This Example
+-------------------
+Now that you have the basics of MapReduce programs down, you can extend this example by:
+
+* Writing a workflow to schedule this MapReduce job every hour and process the last hour's data
```
Comment (abaranau) [inline README.rst]: link to workflow in guide

File: README.rst [hunk 35]
```diff
@@ -0,0 +1,272 @@
...
+* Store the results in a Timeseries data for each run of the MapReduce job
```
Comment (abaranau) [inline README.rst]: to analyze the trends

File: README.rst (hunk 24 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed. "where Apache access logs are ingested to."

File: README.rst (hunk 25 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed

File: README.rst (hunk 26 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed

File: README.rst (hunk 28 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 27 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 29 above)
Comment (sreevatsanraman) [inline README.rst:175]: Not sure if I understand this comment. Is something before 'or smth like that' missing?

File: README.rst (hunk 18 above)
Comment (sreevatsanraman) [inline README.rst]: Done.

File: README.rst (hunk 20 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 21 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 22 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 17 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 30 above)
Comment (sreevatsanraman) [inline README.rst:199]: Fixed.

File: README.rst (hunk 31 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 33 above)
Comment (sreevatsanraman) [inline README.rst]: Fixed.

File: README.rst (hunk 35 above)
Comment (sreevatsanraman) [inline README.rst]: fixed.

File: README.rst [hunk 36]
```diff
@@ -0,0 +1,271 @@
...
+amounts of data in a reliable and fault-tolerant manner. In this guide you will learn how to batch process data using
+MapReduce in the `Cask Data Application Platform (CDAP). <http://cdap.io>`_
+
+What You Will Build
+-------------------
+
+This guide will take you through building a CDAP application that ingests Apache access log events, computes
+top 10 client IPs that sends requests in the last hour and query the results. You will:
+
+* Build `MapReduce<http://docs.cask.co/cdap/current/en/dev-guide.html#mapreduce>`_ job to process Apache access log events
```
Comment (anwar6953) [inline README.rst:14]: To fix the markup rendering, insert a space between 'MapReduce' and '<'

File: README.rst [hunk 37]
```diff
@@ -0,0 +1,271 @@
...
+      // Read events from last 60 minutes as input to the mapper.
+      final long endTime = context.getLogicalStartTime();
+      final long startTime = endTime - TimeUnit.MINUTES.toMillis(60);
+      StreamBatchReadable.useStreamInput(context, "logEvent", startTime, endTime);
+    }
+  }
+
+
+In this example Mapper and Reducer classes are built by implementing
+`Hadoop APIs<http://hadoop.apache.org/docs/r2.3.0/api/org/apache/hadoop/mapreduce/package-summary.html>`_
```
Comment (anwar6953) [inline README.rst:153]: To fix the markup rendering, insert a space between 'APIs' and '<'

Also this:
https://github.com/cdap-guides/cdap-mapreduce-guide/pull/1#discussion_r18937878

**Pull request merged**